### ISSUE TITLE
API Client can list multiple apps via `GET /v3/apps` (with no filters)

### ISSUE THREAD

**Issue by tcdowney** [cloudfoundry/cf-k8s-api#6]
## Blockers
* The CFApp custom resource definition should already exist in https://github.com/cloudfoundry/cf-k8s-controllers/
* https://github.com/cloudfoundry/cf-k8s-api/issues/4
* Implementation with true RBAC blocked on https://docs.google.com/document/d/1M02YxnUfNRZw8-5W9TfWj3jghCPG1M-Pracgk6BSV1I/edit#

## Description
As a client of the CF API, I want to be able to list multiple CFApp resource on Kubernetes via the API shim.

The behavior of this endpoint should match the existing behavior of the [`GET /v3/apps` endpoint](https://v3-apidocs.cloudfoundry.org/version/3.102.0/index.html#list-apps) in CF for VMs as closely as possible -- including the [error cases](https://v3-apidocs.cloudfoundry.org/version/3.102.0/index.html#common-errors).

## Scenarios

**GIVEN** I have the API shim running locally
**THEN** I can make perform the following scenarios

**NOTE** We will handle filtering via query parameters in another story.

---

### Happy Path

**GIVEN** I have applied two `CFApp` CRs to the cluster in different namespaces (TODO: work with @matt-royal for a sample) and know its "guid"
**WHEN** I make the following request

```
curl "https://api.example.org/v3/apps" \
  -X GET \
  -H "Authorization: bearer <some-placeholder-token>" \
  -H "Content-type: application/json"
```

**THEN** I get a response that looks similar to the following (_ensure the fields match what the CR has and make sense_)

```json
HTTP/1.1 200 OK
Content-Type: application/json

  {
    "pagination": {
      "total_results": 2,
      "total_pages": 1,
      "first": {
        "href": "https://api.example.org/v3/apps?page=1"
      },
      "last": {
        "href": "https://api.example.org/v3/apps?page=1"
      },
      "next": null,
      "previous": null
    },
    "resources": [
      {
        "guid": "1cb006ee-fb05-47e1-b541-c34179ddc446",
        "name": "my_app",
        "state": "STARTED",
        "created_at": "2016-03-17T21:41:30Z",
        "updated_at": "2016-03-18T11:32:30Z",
        "lifecycle": {
          "type": "buildpack",
          "data": {
            "buildpacks": ["java_buildpack"],
            "stack": "cflinuxfs3"
          }
        },
        "relationships": {
          "space": {
            "data": {
              "guid": "space-guid-1"
            }
          }
        },
        "links": {
          "self": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446"
          },
          "space": {
            "href": "https://api.example.org/v3/spaces/2f35885d-0c9d-4423-83ad-fd05066f8576"
          },
          "processes": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/processes"
          },
          "packages": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/packages"
          },
          "environment_variables": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/environment_variables"
          },
          "current_droplet": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/droplets/current"
          },
          "droplets": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/droplets"
          },
          "tasks": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/tasks"
          },
          "start": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/actions/start",
            "method": "POST"
          },
          "stop": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/actions/stop",
            "method": "POST"
          },
          "revisions": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/revisions"
          },
          "deployed_revisions": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/revisions/deployed"
          },
          "features": {
            "href": "https://api.example.org/v3/apps/1cb006ee-fb05-47e1-b541-c34179ddc446/features"
          }
        },
        "metadata": {
          "labels": {},
          "annotations": {}
        }
      },
      {
        "guid": "02b4ec9b-94c7-4468-9c23-4e906191a0f8",
        "name": "my_app2",
        "state": "STOPPED",
        "created_at": "1970-01-01T00:00:02Z",
        "updated_at": "2016-06-08T16:41:26Z",
        "lifecycle": {
          "type": "buildpack",
          "data": {
            "buildpacks": ["ruby_buildpack"],
            "stack": "cflinuxfs3"
          }
        },
        "relationships": {
          "space": {
            "data": {
              "guid": "space-guid-2"
            }
          }
        },
        "links": {
          "self": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8"
          },
          "space": {
            "href": "https://api.example.org/v3/spaces/2f35885d-0c9d-4423-83ad-fd05066f8576"
          },
          "processes": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/processes"
          },
          "packages": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/packages"
          },
          "environment_variables": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/environment_variables"
          },
          "current_droplet": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/droplets/current"
          },
          "droplets": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/droplets"
          },
          "tasks": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/tasks"
          },
          "start": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/actions/start",
            "method": "POST"
          },
          "stop": {
            "href": "https://api.example.org/v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/actions/stop",
            "method": "POST"
          },
          "revisions": {
            "href": "https://api.example.org//v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/revisions"
          },
          "deployed_revisions": {
            "href": "https://api.example.org//v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/revisions/deployed"
          },
          "features": {
            "href": "https://api.example.org//v3/apps/02b4ec9b-94c7-4468-9c23-4e906191a0f8/features"
          }
        },
        "metadata": {
          "labels": {},
          "annotations": {}
        }
      }
    ]
  }
```

---

### No Apps exist
**WHEN** I make the following request and no apps exist

```
curl "https://api.example.org/v3/apps" \
  -X GET \
  -H "Authorization: bearer <some-placeholder-token>" \
  -H "Content-type: application/json"
```

**THEN** I get back the following response

```json
HTTP/1.1 200 OK

{
   "pagination": {
      "total_results": 0,
      "total_pages": 1,
      "first": {
         "href": "https://api.coal-carpet.capi.land/v3/apps?page=1"
      },
      "last": {
         "href": "https://api.coal-carpet.capi.land/v3/apps?page=1"
      },
      "next": null,
      "previous": null
   },
   "resources": []
}
```

---

### Dev Notes
* Since Auth(n/z) work is coming in separate story, use a hard-coded `ServiceAccount` for auth for now. For this story the `ServiceAccount` should have the minimal set of permissions to view a `CFApp`. Consider https://github.com/cloudfoundry/cf-crd-explorations/blob/main/config/rbac/app_viewer_role.yaml as an example.
* Docs: https://v3-apidocs.cloudfoundry.org/version/3.102.0/index.html#list-apps
* Cloud Controller code: https://github.com/cloudfoundry/cloud_controller_ng/blob/b241622b890138c6e746f1407cc2cd7aec47cbbf/app/controllers/v3/apps_controller.rb#L37-L63
* Assumes admin credentials for K8s client cache

**Comment by acosta11** [comment]
For an initial development, we're providing the shim with cluster level permissions, but will circle back to user permission delegation when the team decide's how to implement list in such a case. The discussion for that implementation is in https://docs.google.com/document/d/1M02YxnUfNRZw8-5W9TfWj3jghCPG1M-Pracgk6BSV1I/edit#.

**Comment by acosta11** [comment]
Unintentionally closed this issue before the subsequent business logic commit got finished up after linking the initial refactor work we did. Re-opening.

**Comment by Birdrock** [comment]
#76 has been merged.

**Comment by acosta11** [comment]
Forgot to add the routing code here; following up with another commit to actually wrap up the implementation instead of having the api throw 404 page not found.
